libutils/array_map.c

```diff
@@ -86,6 +86,26 @@ bool ArrayMapRemove(ArrayMap *map, const void *key)
     return false;
 }
 
+bool ArrayMapRemoveSoft(ArrayMap *map, const void *key)
+{
+    assert(map != NULL);
+
+    for (int i = 0; i < map->size; ++i)
+    {
+        if (map->equal_fn(map->values[i].key, key))
```

Comment:
I looked up the declarations for the `ArrayMap` and i have a series of interlinked questions.
- Why would each `ArrayMap` "instance" need to have a pointer to the equal, destroy etc functions?
- How is it expected to be used?
- Doesnt the `ArrayMap` have _a_ specific implementation?

I saw the use of `ArrayMapNew` [here](https://github.com/cfengine/libntech/blob/828aa81863f96f243ca3b8a824e1e154f7878df1/libutils/map.c#L107) and [here](https://github.com/cfengine/libntech/blob/828aa81863f96f243ca3b8a824e1e154f7878df1/tests/unit/map_test.c#L531) but im still not sure i _exactly_ get the idea behind giving the map these three functions.

Reply:
Well, the map needs to be able to compare the keys, destroy the keys and destroy data. It's a generic key-value storage so it stores `void` pointers and needs to know how to compare them and how to destroy the data they point to.